streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -38,346 +39,680 @@
+    // NOTE a changelog would only be removed when its corresponding task
+    // is being removed from the thread; otherwise it would stay in this map even after completed
+    private final Map<TopicPartition, ChangelogMetadata> changelogs;
+
+    // the changelog reader only need the main consumer to get committed offsets for source changelog partitions
+    // to update offset limit for standby tasks;
+    private Consumer<byte[], byte[]> mainConsumer;
+
+    void setMainConsumer(final Consumer<byte[], byte[]> consumer) {
+        this.mainConsumer = consumer;
+    }
+
+    public StoreChangelogReader(final StreamsConfig config,
+                                final LogContext logContext,
+                                final Consumer<byte[], byte[]> restoreConsumer,
+                                final StateRestoreListener stateRestoreListener) {
+        this.log = logContext.logger(StoreChangelogReader.class);
+        this.state = ChangelogReaderState.ACTIVE_RESTORING;
+        this.restoreConsumer = restoreConsumer;
+        this.stateRestoreListener = stateRestoreListener;
+
+        // NOTE for restoring active and updating standby we may prefer different poll time
+        // in order to make sure we call the main consumer#poll in time.
+        // TODO: once both of these are moved to a separate thread this may no longer be a concern
+        this.pollTime = Duration.ofMillis(config.getLong(StreamsConfig.POLL_MS_CONFIG));
+
+        this.changelogs = new HashMap<>();
+    }
+
+    private static String recordEndOffset(final Long endOffset) {
+        return endOffset == null ? "UNKNOWN (since it is for standby task)" : endOffset.toString();
+    }
+
+    private boolean restoredToEnd(final ChangelogMetadata metadata) {
+        final Long endOffset = metadata.restoreEndOffset;
+        final Long currentOffset = metadata.stateManager.storeMetadata(metadata.changelogPartition).offset;
+        if (endOffset == null) {
+            // end offset is not initialized meaning that it is from a standby task,
+            // this should never happen since we only call this function for active task in restoring phase
+            throw new IllegalStateException("End offset for changelog " + metadata + " is unknown when deciding " +
+                "if it has completed restoration, this should never happen.");
+        } else if (endOffset == 0) {
+            // this is a special case, meaning there's nothing to be restored since the changelog has no data
+            // OR the changelog is a source topic and there's no committed offset
+            return true;
+        } else if (currentOffset == null) {
+            // current offset is not initialized meaning there's no checkpointed offset,
+            // we would start restoring from beginning and it does not end yet
+            return false;
+        } else {
+            // NOTE there are several corner cases that we need to consider:
```

Comment:
@ableegoldman @vvcephei @cadonna This was a tricky thing that bites us several times due to limit offset / txn markers etc. I hope y current proposal is clean and elegant enough to cover those cases. But please pay extra attention and help me thinking if there are corner cases that can still break it.

Reply:
For standby tasks the end-offset may be not-null if the changelog topic is piggy-backed with the source topic. It is a one-time thing admittedly (most cases the end-offset > committed-offset, and we only take care of log truncation case), but I'm a bit concerned that consolidating them into a single field may be less readable than keeping two fields -- I agree it is a bit verbose, but feel it worth it.

Reply:
> For standby tasks the end-offset may be not-null

In the comments above `ChangelogMetadata.restoreEndOffset`  it says " only for active restoring tasks (for standby changelog it is null)"
I think it's going to get quite confusing to use null/not-null of two fields to encode other information about the `ChangelogMetadata`'s state. It seems like we've already potentially mixed up what it means, and when which one can/will be null.
Instead of consolidating them into a single field, could we just always initialize both and just store a flag `isSource` and maybe `isActive/Standby`? I think it will make the code a _lot_ more clear, and help prevent future bugs. Someone who needs to make a change here should not need to know/remember to null check offsets, and correctly interpret what it means when one is/isn't null

Reply:
Another argument for initializing both is that we will likely _need_ both for KIP-535 and/or KIP-441. I'm not sure how 535 is implementing the offset tracking, but I feel pretty strongly that we should maintain a localized single source of truth for all the offsets. Forcing one of them to be null just to encode some orthogonal information about what kind of task it is is incredibly inflexible and we'll almost surely end up doing another cleanup/refactoring down the line cc @guozhangwang @vvcephei

Reply:
Regarding the consolidation, I've thought about that but there's one tricky thing (not a blocker, just bringing this up):

1. end-offset is only set once and otherwise kept as null: for standby it is never set, for active it is set as min(end-offset, optional(committed-offset)) where committed-offset is required for source-changelog.

2. limit-offset is set multiple times and otherwise kept as null: for source-changelog standby it is initialized to 0 to forbid the standby tasks to be updated at all before we get a first update on the value, for non-source-changelog and active it is always null.

So because for source-changelog standby tasks limit-offset is initialized as 0 while for active tasks the end-offset is initialized as null is the key difference here: of course I can also initialize the end-offset for active as 0 as well but that's not necessary and is a regression.

Regarding the offset tracking, it should be computed as `end-offset minus current-offset` for active and `min(end-offset, optional(limit-offset) minus current-offset` for standby, and current-offset source-of-truth is ProcessorStateManager#StoreMetadata#offset. 

If we want to consolidate them, then one idea would be:

1. for source-changelogs, always initialize limit-offset as 0, otherwise always initialize as INF; and then periodically update the limit-offset for source-changelogs.

2. periodically update the log-end-offset for all changelogs: for active restoring we know the end-offset should not be updated at all.

3. define `end-offset = min(log-end-offset, limit-offset)`. For active and non-source-changelogs it is reduced to `log-end-offset`; however only check for completion for active.

4. the for KIP-535 where we report a tuple of (current-offset, end-offset), current-offset always read from ProcessorStateManager, and end-offset read as:

1) for active processing, just the last offset from RecordCollector.
2) for active restoring, return the `log-end-offset`.
3) for standby, return the `end-offset` calculated above.

To me it seems equally complex compared with the current proposal.

Reply:
Maybe it was a mistake to drag the offset lag computation in to this. We definitely don't need to add any more complexity here :)
But let me put out a more alternative proposal. Note that for both active and standby tasks,
a) regular changelog --> `limitOffset == endOffset` and we restore up to `limitOffset`.
b) source changelog --> `limitOffset <= endOffset` and we restore up to `limitOffset`.
In both cases we actually don't care at all about the current `endOffset` and the only difference between the two cases is that `endOffset` happens to be increasing for source changelogs. But we don't care, we just need to know where to stop. 
Now note, for either kind of changelog we just restore up to the `limitOffset`, where
a) active --> `limitOffset` is constant, we must be careful not to go past it and once `curOffset == limitOffset` we are done (permanently)
b) standby --> `limitOffset` increases, we must be careful not to go past it and once 
`curOffset == limitOffset`
So again, there are more similarities than differences in how we handle them. Here's what I propose then:

1) At no point is anything with "offset" in the name allowed to be null :) They also get initialized immediately and always represent the "correct" value, eg do not use a mock value "0L" to signal we should not restore standbys yet.
2) There is no `endOffset` field, just `limitOffset`. Inside register we just initialize
`limitOffset = exists(committedOffset) ?  min(endOffset, committedOffset) : endOffset;`
which is correct for both regular and source changelogs, and we never need to know which it is.
3) During restoration:
  - first `if (ChangelogMetadata.stateManager.isStandby) ; updateLimitOffset()` 
  - calculate and restore as many buffered records s.t. `curOffset <= limitOffset` holds
  - if we reach the point where `curOffset == limitOffset`*, then for
    standbys: keep remaining buffered records,
    actives: toss any remaining buffered records and mark as completed
  * we would still need all the ugly handling around tx-marker edge cases unfortunately
4)  ???
5) profit

Just to clarify, I don't think we need to block any of your PRs on this refactoring. I would be totally happy to continue merging as is, and go back to clean this up later. Of course who knows what this class will look like once John takes over the branch 😜

Reply:
Thanks @ableegoldman, I like your suggestion here, just to point it out that some complexity as I mentioned above does not reside in the logical workflow itself but around timeout exception handling etc, e.g. when querying committed offset / end offsets if we cannot get it within timeout, the in your step 2) the `endOffset / committedOffset` would not be complete to reliably use. Of course this can be addressed by not transiting the corresponding changelog from `registered` to `initialized` (and hence we do not need the sentinel `null` or `0` values), but at the same time we would also potentially trying to get their `position` to complete the initialization later. Another minor thing worth mentioning is that currently for regular standbys we do not need to update the `limitOffset == endOffset` regularly since we do not complete and we would never read beyond end-offset anyways --- following your idea we can still make the `limitOffset == long.MAX` maybe to bypass the unnecessary end-offset requests.

I will try out your ideas in follow-up PRs.